### PR TITLE
OIDC accounts RUDL API actions

### PR DESCRIPTION
**Pull request by talanknight**

### CODE REVIEW

File: internal/auth/oidc/store/oidc.pb.go [hunk 1]
```diff
@@ -796,63 +796,66 @@ var file_controller_storage_auth_oidc_store_v1_oidc_proto_rawDesc = []byte{
 	0x01, 0x28, 0x0b, 0x32, 0x2a, 0x2e, 0x63, 0x6f, 0x6e, 0x74, 0x72, 0x6f, 0x6c, 0x6c, 0x65, 0x72,
 	0x2e, 0x73, 0x74, 0x6f, 0x72, 0x61, 0x67, 0x65, 0x2e, 0x74, 0x69, 0x6d, 0x65, 0x73, 0x74, 0x61,
 	0x6d, 0x70, 0x2e, 0x76, 0x31, 0x2e, 0x54, 0x69, 0x6d, 0x65, 0x73, 0x74, 0x61, 0x6d, 0x70, 0x52,
-	0x0a, 0x75, 0x70, 0x64, 0x61, 0x74, 0x65, 0x54, 0x69, 0x6d, 0x65, 0x12, 0x12, 0x0a, 0x04, 0x6e,
```
Comment (jimlambrt) [inline internal/auth/oidc/store/oidc.pb.go:799]: what causes this change?

File: internal/auth/subtype.go [hunk 2]
```diff
@@ -42,6 +42,21 @@ var (
...
+	GetCreateTime() *timestamp.Timestamp
+	GetUpdateTime() *timestamp.Timestamp
+	GetName() string
+	GetDescription() string
+	GetAuthMethodId() string
+	GetVersion() uint32
+}
+
+var (
+	_ Account = (*oidc.Account)(nil)
```
Comment (jimlambrt) [inline internal/auth/subtype.go:56]: ty!

File: internal/auth/subtype.go [hunk 3]
```diff
@@ -42,6 +42,21 @@ var (
 	_ AuthMethod = (*password.AuthMethod)(nil)
 )
 
+type Account interface {
```
Comment (jimlambrt) [inline internal/auth/subtype.go:45]: what about the new fields: FullName and Email

File: internal/servers/controller/handlers/accounts/account_service.go [hunk 4]
```diff
@@ -19,17 +21,46 @@ import (
 	"google.golang.org/protobuf/types/known/wrapperspb"
 )
 
+const (
+	// general auth method field names
```
Comment (jimlambrt) [inline internal/servers/controller/handlers/accounts/account_service.go:25]: ty!

File: internal/servers/controller/handlers/accounts/account_service.go [hunk 5]
```diff
@@ -202,26 +240,43 @@ func (s Service) SetPassword(ctx context.Context, req *pbs.SetPasswordRequest) (
...
+		}
+		acct = a
+	default:
+		return nil, handlers.NotFoundErrorf("Unrecognized id.")
 	}
-	return toProto(u)
+	return toProto(acct)
 }
 
 func (s Service) createInRepo(ctx context.Context, authMethodId, scopeId string, item *pb.Account) (*pb.Account, error) {
```
Comment (jimlambrt) [inline internal/servers/controller/handlers/accounts/account_service.go:282]: Will we provide the ability to create OIDC accounts in a future PR?

File: internal/servers/controller/handlers/accounts/account_service.go [hunk 6]
```diff
@@ -322,16 +433,31 @@ func (s Service) deleteFromRepo(ctx context.Context, scopeId, id string) (bool,
 }
 
 func (s Service) listFromRepo(ctx context.Context, authMethodId string) ([]*pb.Account, error) {
-	repo, err := s.repoFn()
+	pwRepo, err := s.pwRepoFn()
 	if err != nil {
 		return nil, err
 	}
-	ul, err := repo.ListAccounts(ctx, authMethodId)
+	pwl, err := pwRepo.ListAccounts(ctx, authMethodId)
```
Comment (jimlambrt) [inline internal/servers/controller/handlers/accounts/account_service.go:440]: why make the db trip if it's not a password auth method?  I'm curious why you don't have a switch here on subtype?

File: internal/servers/controller/handlers/accounts/account_service.go [hunk 7]
```diff
@@ -51,15 +85,19 @@ func init() {
...
 
-	repoFn common.PasswordAuthRepoFactory
+	pwRepoFn   common.PasswordAuthRepoFactory
+	oidcRepoFn common.OidcAuthRepoFactory
 }
 
 // NewService returns a user service which handles user related requests to boundary.
-func NewService(repo common.PasswordAuthRepoFactory) (Service, error) {
-	if repo == nil {
+func NewService(pwRepo common.PasswordAuthRepoFactory, oidcRepo common.OidcAuthRepoFactory) (Service, error) {
```
Comment (jimlambrt) [inline internal/servers/controller/handlers/accounts/account_service.go:93]: Mind adding a test for invalid parameters?

File: internal/servers/controller/handlers/accounts/account_service.go [hunk 8]
```diff
@@ -202,26 +240,43 @@ func (s Service) SetPassword(ctx context.Context, req *pbs.SetPasswordRequest) (
 		return nil, err
 	}
 	u.Scope = authResults.Scope
-	u.AuthorizedActions = authResults.FetchActionSetForId(ctx, u.Id, IdActions).Strings()
+	u.AuthorizedActions = authResults.FetchActionSetForId(ctx, u.Id, IdActions[auth.SubtypeFromId(u.Id)]).Strings()
 	return &pbs.SetPasswordResponse{Item: u}, nil
 }
 
 func (s Service) getFromRepo(ctx context.Context, id string) (*pb.Account, error) {
```
Comment (jimlambrt) [inline internal/servers/controller/handlers/accounts/account_service.go:247]: mind adding tests for IsNotFound for both subtypes?

File: internal/servers/controller/handlers/accounts/account_service.go [hunk 9]
```diff
@@ -303,15 +344,85 @@ func (s Service) updateInRepo(ctx context.Context, scopeId, authMethId, id strin
...
+	case auth.OidcSubtype:
+		repo, iErr := s.oidcRepoFn()
+		if iErr != nil {
+			return false, iErr
+		}
+		rows, err = repo.DeleteAccount(ctx, scopeId, id)
 	}
-	rows, err := repo.DeleteAccount(ctx, scopeId, id)
 	if err != nil {
 		if errors.IsNotFoundError(err) {
```
Comment (jimlambrt) [inline internal/servers/controller/handlers/accounts/account_service.go:427]: mind adding a test for this condition?

File: internal/servers/controller/handlers/accounts/account_service.go [hunk 10]
```diff
@@ -342,7 +468,7 @@ func (s Service) listFromRepo(ctx context.Context, authMethodId string) ([]*pb.A
 }
 
 func (s Service) changePasswordInRepo(ctx context.Context, scopeId, id string, version uint32, currentPassword, newPassword string) (*pb.Account, error) {
-	repo, err := s.repoFn()
+	repo, err := s.pwRepoFn()
 	if err != nil {
 		return nil, err
 	}
```
Comment (jimlambrt) [inline internal/servers/controller/handlers/accounts/account_service.go:474]: mind adding tests for NotFound and password too short conditions?

File: internal/servers/controller/handlers/accounts/account_service.go [hunk 11]
```diff
@@ -367,7 +493,7 @@ func (s Service) changePasswordInRepo(ctx context.Context, scopeId, id string, v
 }
 
 func (s Service) setPasswordInRepo(ctx context.Context, scopeId, id string, version uint32, pw string) (*pb.Account, error) {
-	repo, err := s.repoFn()
+	repo, err := s.pwRepoFn()
 	if err != nil {
 		return nil, err
 	}
```
Comment (jimlambrt) [inline internal/servers/controller/handlers/accounts/account_service.go:499]: mind adding tests for NotFound and password too short conditions?

File: internal/servers/controller/handlers/accounts/account_service.go [hunk 12]
```diff
@@ -385,132 +511,250 @@ func (s Service) setPasswordInRepo(ctx context.Context, scopeId, id string, vers
...
 func validateGetRequest(req *pbs.GetAccountRequest) error {
-	return handlers.ValidateGetRequest(handlers.NoopValidatorFn, req, password.AccountPrefix)
+	const op = "account_service.validateGetRequest"
+	if req == nil {
+		return errors.New(errors.InvalidParameter, op, "nil request")
+	}
+	return handlers.ValidateGetRequest(handlers.NoopValidatorFn, req, password.AccountPrefix, oidc.AccountPrefix)
 }
 
 func validateCreateRequest(req *pbs.CreateAccountRequest) error {
```
Comment (jimlambrt) [inline internal/servers/controller/handlers/accounts/account_service.go:670]: would you mind adding test coverage for this func?

File: internal/servers/controller/handlers/accounts/account_service.go [hunk 13]
```diff
@@ -385,132 +511,250 @@ func (s Service) setPasswordInRepo(ctx context.Context, scopeId, id string, vers
...
+			badFields[authMethodIdField] = "Unable to create accounts for this auth method type."
 		default:
-			badFields["auth_method_id"] = "Unknown auth method type from ID."
+			badFields[authMethodIdField] = "Unknown auth method type from ID."
 		}
 		return badFields
 	})
 }
 
 func validateUpdateRequest(req *pbs.UpdateAccountRequest) error {
```
Comment (jimlambrt) [inline internal/servers/controller/handlers/accounts/account_service.go:702]: mind adding additional test coverage for this func?

File: internal/servers/controller/handlers/authmethods/authmethod_service.go [hunk 14]
```diff
@@ -71,8 +71,6 @@ var (
 			action.Read,
 			action.Update,
 			action.Delete,
-			action.ChangePassword,
```
Comment (jimlambrt) [inline internal/servers/controller/handlers/authmethods/authmethod_service.go:74]: I know it's not really in scope for this PR... would you mind adding better coverage for some of these functions?  There's a couple that appear to have no coverage.  ty!

**Review by jimlambrt — CHANGES_REQUESTED** [review]
Great work!  Just a few minor questions and suggestions.

File: internal/servers/controller/handlers/accounts/account_service.go (hunk 8 above)
Comment (talanknight) [reply internal/servers/controller/handlers/accounts/account_service.go:247]: Added

**Review by talanknight — COMMENTED** [review]
Thank you!

**Review by jimlambrt — APPROVED** [review]
ty!

**Pull request merged**